hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -1537,6 +1640,277 @@ public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
 		}
 	}
 
+	protected class AllowancePrecompile extends ERCReadOnlyAbstractPrecompile {
+		private TokenAllowanceWrapper allowanceWrapper;
+
+		public AllowancePrecompile(final TokenID tokenId) {
+			super(tokenId);
+		}
+
+		@Override
+		public TransactionBody.Builder body(final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+			final var nestedInput = input.slice(24);
+			allowanceWrapper = decoder.decodeTokenAllowance(nestedInput, aliasResolver);
+
+			return super.body(input, aliasResolver);
+		}
+
+		@Override
+		public Bytes getSuccessResultFor(final ExpirableTxnRecord.Builder childRecord) {
+			final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger = ledgers.accounts();
+			final var allowances = (TreeMap<FcTokenAllowanceId, Long>) accountsLedger.get(allowanceWrapper.owner(), FUNGIBLE_TOKEN_ALLOWANCES);
+			long value = 0;
+			final var fcTokenAllowanceId =
+					FcTokenAllowanceId.from(EntityNum.fromTokenId(tokenId), EntityNum.fromAccountId(allowanceWrapper.spender()));
+			if (allowances.containsKey(fcTokenAllowanceId)) {
+				value = allowances.get(fcTokenAllowanceId);
+			}
+			return encoder.encodeAllowance(value);
+		}
+	}
+
+	protected class ApprovePrecompile implements Precompile {
+		protected ApproveWrapper approveOp;
+		protected TokenID token;
+		private final boolean isFungible;
+
+		public ApprovePrecompile(final TokenID token, final boolean isFungible) {
+			this.token = token;
+			this.isFungible = isFungible;
+		}
+
+		@Override
+		public TransactionBody.Builder body(final Bytes input, final UnaryOperator<byte[]> aliasResolver) {
+			final var accountId = EntityIdUtils.accountIdFromEvmAddress(senderAddress);
+			final var fungibleTokenAllowances = (Map<FcTokenAllowanceId, Long>) ledgers.accounts().get(accountId, AccountProperty.FUNGIBLE_TOKEN_ALLOWANCES);
+
+			final var nestedInput = input.slice(24);
+			approveOp = decoder.decodeTokenApprove(nestedInput, token, isFungible, aliasResolver);
+
+			if (isFungible) {
+				long value = 0;
+				final var fcTokenAllowanceId =
+						FcTokenAllowanceId.from(EntityNum.fromTokenId(token), EntityNum.fromAccountId(approveOp.spender()));
+				if (fungibleTokenAllowances.containsKey(fcTokenAllowanceId)) {
+					value = fungibleTokenAllowances.get(fcTokenAllowanceId);
+				}
+				return syntheticTxnFactory.createApproveAllowance(approveOp.withAdjustment(approveOp.amount().subtract(BigInteger.valueOf(value))));
```

Comment:
This is wrong. We had to calculate adjustments when we used to submit `CryptoAdjustApprovalTransactions`, now we simply need to create a synthetic `CryptoApproveAllowanceTransactionBody` with the value received as input arg, and the old value will be overridden with the new value. No need to do any calculations.